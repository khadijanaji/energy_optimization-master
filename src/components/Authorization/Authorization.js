/* eslint-disable react/prop-types,react/no-unused-prop-types */
// @flow

import * as React from "react";
import type {InfoUserPropsType} from "../../storeTypes/LoginForm";
import {get} from "lodash";

export type AppPropsType = {
  redirectTo: Function,
  isUserConnected: boolean,
  children: React.Node,
  user: InfoUserPropsType,
  params: {
    entity: string,
    domain: string
  }
};

const userHasAccess = (domainRequest: string, entityRequest: string, user: InfoUserPropsType): boolean => {
  const userDomains = get(user, "roles", []);
  const userEntities = get(user, "factories", []);
  return (!domainRequest || userDomains.includes(domainRequest.toUpperCase())) &&
    (!entityRequest || userEntities.includes(entityRequest.toUpperCase()));
};

const redirectIfNotAuthorized = (props: AppPropsType) => {
  const {user, redirectTo, isUserConnected, "params": {entity = "", domain = ""}} = props;
  if (!isUserConnected) {
    redirectTo("/login");
    return;
  }
  if (!userHasAccess(domain, entity, user)) {
    redirectTo("/");
  }
};


class Authorization extends React.Component<AppPropsType> {
  componentDidMount () {
    redirectIfNotAuthorized(this.props);
  }

  componentWillReceiveProps (nextProps: AppPropsType) {
    const {children, "params": {entity, domain}} = this.props;
    const {"children": nextChildren, "params": {"domain": nextDomain, "entity": nextEntity}} = nextProps;
    if (nextChildren !== children || entity !== nextEntity || nextDomain !== domain) {
      redirectIfNotAuthorized(nextProps);
    }
  }

  render () {
    const {children} = this.props;

    return (<React.Fragment>
      {children}
    </React.Fragment>);
  }
}

export default Authorization;
