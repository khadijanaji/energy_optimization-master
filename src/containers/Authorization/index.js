// @flow

import {connect} from "react-redux";
import {push} from "react-router-redux";

import App from "../../components/Authorization";
import appSelector from "./selectors";

const mapsStateToProps = appSelector;

const mapDispatchToProps = {"redirectTo": push};

export default connect(mapsStateToProps, mapDispatchToProps)(App);
