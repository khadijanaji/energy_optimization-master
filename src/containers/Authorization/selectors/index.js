// @flow

import {tokenSelector} from "../../LoginForm/selectors";
import {get} from "lodash";

const appSelector = (state: any): any => {
  const token = tokenSelector(state);
  const user = get(state, [
    "login",
    "infoUser"
  ], {});
  return {
    "isUserConnected": Boolean(token),
    user
  };
};

export default appSelector;
