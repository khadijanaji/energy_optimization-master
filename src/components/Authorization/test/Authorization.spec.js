/* eslint-disable array-element-newline */
/* globals describe it expect */

import {shallow, mount} from "enzyme";
import React from "react";
import sinon from "sinon";
import Authorization from "../";

let props = {};
let redirectSpyFunction = () => ({});
const userRoles = ["SAP", "DAP"];
const userFactory = "JFC2";

describe("Component", () => {
  describe("Authorization Render", () => {
    beforeAll(() => {
      redirectSpyFunction = sinon.spy();
      props = {
        "redirectTo": redirectSpyFunction,
        "isUserConnected": true,
        "children": (<div>{"Some Page"}</div>),
        "user": {
          "email": "mail@mail.ma",
          "name": "random user",
          "roles": userRoles,
          "factories": [userFactory]
        },
        "params": {
          "entity": "JFC2",
          "domain": "SAP"
        }
      };
    });
    it("Should render children", () => {
      const wrapper = shallow(<Authorization {...props} />);
      expect(wrapper).toMatchSnapshot();
    });

    it("Updating isUserConnected to false should redirect to login", () => {
      props.isUserConnected = false;
      mount(<Authorization {...props} />);
      expect(redirectSpyFunction.calledWith("/login")).toEqual(true);
    });

    it("Updating domain to value in user roles should not redirect", () => {
      const component = mount(<Authorization {...props} />);
      component.setProps({
        "params": {
          "domain": userRoles[1],
          "entity": userFactory
        }
      });
      expect(redirectSpyFunction.calledWith("/")).toEqual(false);
    });

    it("Should not trigger auth check if params unchanged", () => {
      const component = mount(<Authorization {...props} />);
      component.setProps({
        "params": {
          "domain": userRoles[0],
          "entity": userFactory
        }
      });
      expect(redirectSpyFunction.calledWith("/")).toEqual(false);
    });

    it("Updating domain to value NOT in user roles should redirect", () => {
      const ownProps = {
        "redirectTo": sinon.spy(),
        "isUserConnected": true,
        "children": (<div>{"Some Page"}</div>),
        "user": {
          "email": "mail@mail.ma",
          "name": "random user",
          "roles": userRoles,
          "factories": [userFactory]
        },
        "params": {
          "entity": "JFC2",
          "domain": "SAP"
        }
      };
      const component = mount(<Authorization {...ownProps} />);
      component.setProps({
        "params": {
          "domain": "RandomValue",
          "entity": "RandomValue"
        }
      });
      expect(ownProps.redirectTo.calledWith("/")).toEqual(true);
    });
    it("User has access to Home Screen", () => {
      const component = mount(<Authorization {...props} />);
      component.setProps({
        "params": {
          "domain": undefined,
          "entity": undefined
        }
      });
      expect(redirectSpyFunction.calledWith("/")).toEqual(false);
    });
  });
});
