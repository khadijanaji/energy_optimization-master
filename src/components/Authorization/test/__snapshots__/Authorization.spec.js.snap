// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Authorization Render Should render children 1`] = `
<Fragment>
  <div>
    Some Page
  </div>
</Fragment>
`;
